experimental/uefi/src/main.rs

```diff
@@ -0,0 +1,66 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#![no_main]
+#![no_std]
+#![feature(abi_efiapi)]
+#![feature(custom_test_frameworks)]
+#![test_runner(crate::test_runner)]
+#![reexport_test_harness_main = "test_main"]
+
+use uefi::{prelude::*, table::runtime::ResetType, ResultExt};
+
+#[entry]
+fn _start(_handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
+    uefi_services::init(&mut system_table).unwrap_success();
+
+    let status = if cfg!(test) {
+        #[cfg(test)]
+        test_main();
+        Status::SUCCESS
+    } else {
+        main(_handle, &mut system_table)
+    };
+
+    system_table
+        .runtime_services()
+        .reset(ResetType::Shutdown, status, None);
+}
+
+fn main(_handle: Handle, system_table: &mut SystemTable<Boot>) -> Status {
+    use core::fmt::Write;
+
+    let status = writeln!(system_table.stdout(), "Hello World!");
+
+    match status {
```

Comment:
Can this be expressed more "functionally"? e.g. `status.map(|_| SUCCESS).or(DEVICE_ERROR)` or something like that

Reply:
I could do `status.map_or_else(|_| DEVICE_ERROR, |_| SUCCESS)` but quite frankly I think that's more opaque and hard to read than that simple match.

Reply:
I agree, `map_or_else` is quite unreadable unless you know exactly which way around the args are.

I think the version of what I suggested that *actually* works would be `status.map(|_| SUCCESS).unwrap_or(ERROR)` (which I personally think reads like idiomatic Rust).

Reply:
Yeah, that's much better. Done.